### PR TITLE
console: add lxc.console.buffer.size, lxc.console.buffer.logfile, lxc.console.rotate

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by hallyn** [comment]
My last question would be - do we want to call it .1 instead of .old?  Maybe make the lxc.console.buffer.rotate value be an integer, maximum number to rotate to?  That's more work and probably not needed, but if we start calling it . old now then we can't as easily switch to .N later.

**Comment by brauner** [comment]
@hallyn, I was just following LXD's style but sure.

**Comment by brauner** [comment]
> My last question would be - do we want to call it .1 instead of .old? Maybe make the lxc.console.buffer.rotate value be an integer, maximum number to rotate to?

Internally it is already an integer. But for now I would like to only allow it to be set to `0` or `1`.

**Comment by brauner** [comment]
@hallyn, updated.

File: src/lxc/commands.c [hunk 1]
```diff
@@ -1086,6 +1090,31 @@ static int lxc_cmd_console_log_callback(int fd, struct lxc_cmd_req *req,
...
+		/* rotate the console log file */
+		if (!console->log_path || console->log_rotate == 0)
+			goto out;
+
+		/* be very certain things are kosher */
+		rsp.ret = -EBADF;
+		if (console->log_fd < 0)
+			goto out;
+
+		len = strlen(console->log_path) + sizeof(".old");
```
Comment (hallyn) [inline src/lxc/commands.c]: sizeof(".1")? :)

File: src/lxc/confile.c [hunk 2]
```diff
@@ -1790,6 +1792,23 @@ static int set_config_console_path(const char *key, const char *value,
...
+{
+	if (lxc_config_value_empty(value)) {
+		lxc_conf->console.log_rotate = 0;
+		return 0;
+	}
+
+	if (lxc_safe_uint(value, &lxc_conf->console.log_rotate) < 0)
+		return -1;
+
+	if (lxc_conf->console.log_rotate > 1)
```
Comment (hallyn) [inline src/lxc/confile.c]: Can you add a LOG or ERROR message here?  I probably would have just set log_rotate = 1 if it was > 1, but that's misleading so what you're doing is probably better.  But you should explain to people why they're getting an error.

File: doc/lxc.container.conf.sgml.in [hunk 3]
```diff
@@ -707,6 +707,26 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
...
 
+        <varlistentry>
+          <term>
+            <option>lxc.console.buffer.logfile</option>
+          </term>
+          <listitem>
+            <para>
+            Setting this option instructs liblxc to write the in-memory
+            ringbuffer to disk. For performance reasons liblxc will only write
+            the in-memory ringbuffer to disk when an appropriate API request is
```
Comment (hallyn) [inline doc/lxc.container.conf.sgml.in]: What is an appropriate API request? :)

File: src/lxc/commands.c (hunk 1 above)
Comment (brauner) [reply src/lxc/commands.c]: dammit

File: src/lxc/confile.c (hunk 2 above)
Comment (brauner) [reply src/lxc/confile.c]: Right. Will update in a bit.

File: doc/lxc.container.conf.sgml.in (hunk 3 above)
Comment (brauner) [reply doc/lxc.container.conf.sgml.in]: reworded to "when requested"

**Pull request merged**